Test/PcapDeviceTest.cs

```diff
@@ -51,12 +60,82 @@ public void DeviceProperties([PcapDevices] DeviceFixture fixture)
             Assert.IsNotNull(pcapIf.GatewayAddresses);
             Assert.IsNotNull(pcapIf.Addresses);
 
+            var resolution = device.TimestampPrecision;
+            if(targetResolution != null)
+            {
+                // confirm the resolution was set
+                if(targetResolution != resolution)
+                {
+                    Console.WriteLine("Error: {0} expected {1} but was {2}", device.Name, targetResolution, resolution);
+                } else
+                {
+                    Console.WriteLine("Pass: {0} expected {1} but was {2}", device.Name, targetResolution, resolution);
+                }
+
+                // Note: This doesn't not appear to work for any platform so it is
+                // disabled for now
+                //                Assert.AreEqual(targetResolution, resolution);
+            }
+
+            Assert.IsNotNull(device.TimestampsSupported);
+
             if (pcapIf.MacAddress != null)
             {
                 Assert.That(pcapIf.MacAddress.GetAddressBytes(), Has.Length.EqualTo(6));
             }
         }
 
+        /// <summary>
+        /// Note: This tests the default timestamp precision as set by libpcap internally
+        /// </summary>
+        /// <param name="fixture"></param>
+        [Test]
+        public void DeviceOpen([PcapDevices] DeviceFixture fixture)
+        {
+            using var device = (PcapDevice)fixture.GetDevice();
+
+            device.Open();
+            DeviceTest(device, null);
+        }
+
+        [Test]
+        public void DeviceOpenWithTimestampPrecision([PcapDevices] DeviceFixture fixture)
+        {
+            using var device = (PcapDevice)fixture.GetDevice();
+
+            foreach (TimestampResolution resolution in Enum.GetValues(typeof(TimestampResolution)))
+            {
+                var configuration = new DeviceConfiguration();
+                configuration.TimestampResolution = resolution;
+                device.Open(configuration);
+                DeviceTest(device, resolution);
+            }
+        }
+
+        /// <summary>
+        /// Test specifying the timestamp type when opening devices
+        /// </summary>
+        /// <param name="fixture"></param>
+        [Test]
+        public void DeviceOpenWithTimestampType([PcapDevices] DeviceFixture fixture)
+        {
+            using var device = (PcapDevice)fixture.GetDevice();
+
+            // open the device to retrieve the supported timestamps
+            device.Open();
+            var timestampTypes = device.TimestampsSupported;
+            device.Close();
```

Comment:
Having to open and close device to get list of supported timestamps sounds counter intuitive.

Reply:
That's how they roll over there in libpcap land, see the function we use to retrieve the list, https://www.tcpdump.org/manpages/pcap_list_tstamp_types.3pcap.html

Reply:
we can hide this complexity by creating a temporary handle internally.

Reply:
I don't think there is anything wrong with opening and closing devices or just creating the handle and using that. Yeah let me fix.

Reply:
@kayoub5 since we can retrieve timestamps for any kind of device I'm not sure how this maps to things like capture reader and capture writer devices. At the PcapDevice level we don't know what kind of device it is. I did add a check to throw if not open though, which is necessary.

Reply:
what about adding this api to live device class instead of the base class here?

Reply:
Then we'd leave out the offline devices, the file reader and writer, which can also have their own possible timestamps. I don't see a good path forward yet but I'm guessing there is one.

Reply:
file device can have a timestamp resolution, but not a timestamp type

Reply:
ooh a very good spot, yeah they have their own functions for specifying the timestamp resolution.

Let me move that into the live device.


For capture reader and writer it seems like we could either add the option to the constructor or switch to having the user call Open() on the devices. The former is simpler for the user, the latter is more uniform with opening capture devices. I'm curious which one you think feels better, or do you have another approach?

Reply:
If timestamp is the ONLY option available for files then simply adding it to the constructor would be simpler, if there are more options that could be configured, the `Open()` method would make more sense then.

A third option would be to make `Open()` static factory method as per the proposal in #201

Reply:
@kayoub5 for the static factory approach I presume internally we'd preserve the classes and constructors and return the objects from the appropriate static factory method? It looks more tidy with Pcap.Openxxx() but how does it reduce effort/code/simplify syntax for the user over the current approach, now that we've paired down the Open() calls?